File: site/backend/task_service/client/pkg/app/task/model/task.go
package model

import (
	"backend/client/pkg/app/organization/model"
	"backend/client/pkg/app/paginate"
	"backend/client/pkg/app/task/data"
	"context"
	"time"
)

type TaskModel struct {
	ID                uint      `gorm:"column:id;type:SERIAL;primaryKey;autoIncrement" json:"id"`
	OrganizationID    uint      `gorm:"column:organization_id;type:INTEGER;not null" json:"organization_id"`
	Name              string    `gorm:"column:name;type:VARCHAR(255);not null" json:"name"`
	TypeID            uint      `gorm:"column:type_id;type:INTEGER;not null;index" json:"type_id"`
	Description       string    `gorm:"column:description;type:TEXT;not null" json:"description"`
	Location          string    `gorm:"column:location;type:VARCHAR(255);not null" json:"location"`
	TaskDate          time.Time `gorm:"column:task_date;type:TIMESTAMP;not null" json:"task_date"`
	ParticipantsCount *int      `gorm:"column:participants_count;type:INTEGER" json:"participants_count"`
	MaxScore          *int      `gorm:"column:max_score;type:INTEGER" json:"max_score"`
	StatusID          uint      `gorm:"column:status_id;type:INTEGER;default:1;index" json:"status_id"`
	CreatedAt         time.Time `gorm:"column:created_at;type:TIMESTAMP;default:CURRENT_TIMESTAMP;autoCreateTime" json:"created_at"`
	UpdatedAt         time.Time `gorm:"column:updated_at;type:TIMESTAMP;default:CURRENT_TIMESTAMP;autoUpdateTime" json:"updated_at"`
}

func (TaskModel) TableName() string {
	return "task"
}

type TaskUser struct {
	ID            uint `gorm:"column:id;type:SERIAL;primaryKey;autoIncrement"`
	TaskID        uint `gorm:"column:task_id;type:INTEGER;not null;index"`
	UserID        uint `gorm:"column:user_id;type:INTEGER;not null;index"`
	IsCoordinator bool `gorm:"column:is_coordinator;type:BOOLEAN;default:false"`
}

func (TaskUser) TableName() string {
	return "task_user"
}

type CategoryModel struct {
	ID   uint   `gorm:"column:id;type:SERIAL;primaryKey;autoIncrement"`
	Name string `gorm:"column:name;type:VARCHAR(100);unique;not null"`
}

func (CategoryModel) TableName() string {
	return "category"
}

type TaskCategory struct {
	ID         uint `gorm:"column:id;type:SERIAL;primaryKey;autoIncrement" json:"id"`
	TaskID     uint `gorm:"column:task_id;type:INTEGER;not null;index" json:"task_id"`
	CategoryID uint `gorm:"column:category_id;type:INTEGER;not null;index" json:"category_id"`
}

func (TaskCategory) TableName() string {
	return "task_category"
}

type TaskTypeModel struct {
	ID   uint    `gorm:"column:id;type:SERIAL;primaryKey;autoIncrement" json:"id"`
	Name *string `gorm:"column:name;type:VARCHAR(255)" json:"name"`
}

func (TaskTypeModel) TableName() string {
	return "task_type"
}

type TaskStatusModel struct {
	ID   uint    `gorm:"column:id;type:SERIAL;primaryKey;autoIncrement" json:"id"`
	Name *string `gorm:"column:name;type:VARCHAR(255)" json:"name"`
}

func (TaskStatusModel) TableName() string {
	return "task_status"
}

type TaskReadRepositoryInterface interface {
	Get(ctx context.Context, id uint) (*TaskModel, error)
	GetAll(
		ctx context.Context,
		pagination *paginate.Pagination,
		user uint,
		organizations []model.OrganizationModel,
	) (*paginate.Pagination, error)
}

type TaskRepositoryInterface interface {
	TaskReadRepositoryInterface
	Delete(ctx context.Context, id uint) error
	Update(ctx context.Context, task *data.UpdateTask) error
	Create(ctx context.Context, task *data.CreateTask) (uint, error)
}

type TaskUserReadRepositoryInterface interface {
	GetUsers(ctx context.Context, taskID uint, pagination *paginate.Pagination, isCoordinators *bool) (*paginate.Pagination, error)
}

type TaskUserRepositoryInterface interface {
	TaskUserReadRepositoryInterface
	Add(ctx context.Context, userID, taskID uint, isCoordinator bool) error
	Delete(ctx context.Context, userID, taskID uint) error
}

File: site/backend-yii2/common/models/Task.php
<?php

namespace common\models;

use Yii;
use yii\db\ActiveRecord;

/**
 * This is the model class for table "{{%task}}".
 *
 * @property int $id
 * @property int $organization_id
 * @property string $name
 * @property int $type_id
 * @property string $description
 * @property string $location
 * @property string $task_date
 * @property int|null $participants_count
 * @property int|null $max_score
 * @property int $status_id
 * @property string $created_at
 * @property string $updated_at
 * @property integer $is_deleted Удалено 1/0
 *
 * @property Organization $organization
 * @property TaskType $type
 * @property TaskStatus $status
 * @property TaskUser[]|null $taskUsers
 * @property User[]|null $usersViaTaskUser
 * @property User[]|null $coordinators
 * @property User[]|null $participants
 * @property TaskCategory[] $taskCategories
 * @property Category[] $categories
 */
class Task extends ActiveRecord
{
    public static function tableName()
    {
        return '{{%task}}';
    }

    public function rules()
    {
        return [
            [['organization_id', 'name', 'type_id', 'description', 'location', 'task_date'], 'required'],
            [['organization_id', 'type_id', 'participants_count', 'max_score', 'status_id', 'is_deleted'], 'integer'],
            [['task_date', 'created_at', 'updated_at'], 'safe'],
            [['name', 'description', 'location'], 'string', 'max' => 255],
        ];
    }

    public function attributeLabels()
    {
        return [
            'id' => Yii::t('app', 'ID'),
            'organization_id' => Yii::t('app', 'Organization ID'),
            'name' => Yii::t('app', 'Name'),
            'type_id' => Yii::t('app', 'Type ID'),
            'description' => Yii::t('app', 'Description'),
            'location' => Yii::t('app', 'Location'),
            'task_date' => Yii::t('app', 'Task Date'),
            'participants_count' => Yii::t('app', 'Participants Count'),
            'max_score' => Yii::t('app', 'Max Score'),
            'status_id' => Yii::t('app', 'Status ID'),
            'created_at' => Yii::t('app', 'Created At'),
            'updated_at' => Yii::t('app', 'Updated At'),
        ];
    }

    public function getOrganization()
    {
        return $this->hasOne(Organization::class, ['id' => 'organization_id']);
    }

    public function getType()
    {
        return $this->hasOne(TaskType::class, ['id' => 'type_id']);
    }

    public function getStatus()
    {
        return $this->hasOne(TaskStatus::class, ['id' => 'status_id']);
    }

    public function getTaskUsers()
    {
        return $this->hasMany(TaskUser::class, ['task_id' => 'id']);
    }

    public function getUsersViaTaskUser()
    {
        return $this->hasMany(User::class, ['id' => 'user_id'])->via('taskUsers');
    }

    public function getCoordinators()
    {
        return $this->hasMany(User::class, ['id' => 'user_id'])
            ->via('taskUsers', function ($query) {
                $query->andWhere(['is_coordinator' => 1]);
            });
    }

    public function getParticipants()
    {
        return $this->hasMany(User::class, ['id' => 'user_id'])
            ->via('taskUsers', function ($query) {
                $query->andWhere(['is_coordinator' => 0]);
            });
    }

    public function getTaskCategories()
    {
        return $this->hasMany(TaskCategory::class, ['task_id' => 'id']);
    }

    public function getCategories()
    {
        return $this->hasMany(Category::class, ['id' => 'category_id'])
            ->via('taskCategories');
    }
}
